### ISSUE TITLE
Optimizations (ifneedbe)

### ISSUE THREAD

**Issue by oberien** [oberien/pundoc#73]
This issue is a collection of possible optimizations which might or might not be needed. Benchmarks can be seen in the comments.

- [ ] frontend: Use state-machine. Currently we basically perform a recursive-descend on the events. When we find an event with inner elements (`Event::Start`), we convert the whole event until the corresponding `Event::End` and write all converted events into a buffer. We then remove elements from the buffer until it is empty before parsing the next event.  
This could be replaced with a state-machine, which only converts a single event at a time instead of buffering stuff. This will be quite more complex though and might not be worth it considering performance gain versus mental overhead.
- [ ] everywhere: Don't use `dyn Write`. Currently we use `dyn Write` everywhere, because there might be a lot of different types involved, which can be written to (`Stdout`, `File`, `Vec<u8>` for intermediate caching in CodeGenUnits). We could create a custom enum wrapper instead, which might enable further optimizations (which are not possible via vtables).
- [ ] concurrency: Currently only a single thread is used. There are two ways we can achieve multithreading.
    * CompilationUnit multithreading: This should be done after #71. For every compilation unit (i.e. markdown file encountered), we start a new thread which compiles that file. After every unit has finished compiling, we merge them into a single file (just as we need to do for #71 anyways).
    * Internal multithreading: We find a way to multithread the compilation of a single compilation unit. I don't really have any idea how to achieve that yet, but if it's needed, we can take a closer look.
- [ ] String types: Currently we use `Cow<'a, str>`, but may want to be more careful / switch to the `SmortStr` of `pulldown-cmark` which can hold a character inline. We may also want to think about not using the `Concat` module, which may create large strings.
- [x] BufWriter: https://github.com/oberien/heradoc/issues/73#issuecomment-479427795 Required (#74)
- [ ] ThinLTO: https://github.com/oberien/heradoc/issues/73#issuecomment-479430628 Only worth it for actual release artifacts, not for benchmarking / testing.

Template
````
Benchmark $WHAT on 1300× [crdt-details.md] (96MB):

tldr: 

`perf stat ./target/release/heradoc bench.md -o bench.tex`
```

```

`perf record`
```

```

[crdt-details.md]: https://github.com/xi-editor/xi-editor/blob/564cbde392a2ba8d78a069bcf3dc6aa6a14c2aae/docs/docs/crdt-details.md
````

**Comment by oberien** [comment]
Benchmark before any optimizations on 1300× [crdt-details.md] (96MB):

tldr: >50% of time is spent writing to disk. We should use `BufWriter`.

`perf stat ./target/release/heradoc bench.md -o bench.tex`
```
 Performance counter stats for './target/release/heradoc bench.md -o bench.tex':

          7,697.16 msec task-clock:u              #    1.000 CPUs utilized          
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
            71,526      page-faults:u             #    0.009 M/sec                  
    12,329,993,944      cycles:u                  #    1.602 GHz                    
    19,904,249,535      instructions:u            #    1.61  insn per cycle         
     4,043,431,782      branches:u                #  525.315 M/sec                  
        77,695,697      branch-misses:u           #    1.92% of all branches        

       7.700925988 seconds time elapsed

       3.556403000 seconds user
       4.128888000 seconds sys
```

`perf record -g`
```
+   57.03%    57.03%  heradoc          [unknown]           [k] 0xffffffff8e600163
+   56.74%     0.39%  heradoc          libpthread-2.28.so  [.] __libc_write
+    9.98%     9.59%  heradoc          libc-2.28.so        [.] __memmove_avx_unaligned_erms
+    9.61%     0.35%  heradoc          heradoc             [.] heradoc::generator::iter::Iter::peek
+    9.22%     0.00%  heradoc          [unknown]           [k] 0000000000000000
+    8.70%     0.00%  heradoc          [unknown]           [k] 0x206564756c636e69
+    8.68%     0.00%  heradoc          [unknown]           [k] 0x7474747865745c7b
+    7.05%     1.33%  heradoc          heradoc             [.] heradoc::generator::iter::Iter::next
+    5.76%     5.25%  heradoc          heradoc             [.] <heradoc::backend::latex::simple::TextGen as heradoc::backend::MediumCodeGenUnit<alloc::borrow::Cow<str>>>::gen
```
[crdt-details.md]: https://github.com/xi-editor/xi-editor/blob/564cbde392a2ba8d78a069bcf3dc6aa6a14c2aae/docs/docs/crdt-details.md

**Comment by oberien** [comment]
Benchmark with BufWriter on 1300× [crdt-details.md] (96MB):

tldr: >50% performance gain just for using a `BufWriter`. Problem now seems to be too many memmoves. Most of them are produced by `generator::iter::Iter::peek` (which makes sense as it's the main function converting events). Also the `TextGen` with its char replacing uses over 14% of time.

`perf stat ./target/release/heradoc bench.md -o bench.tex`
```
 Performance counter stats for './target/release/heradoc bench.md -o bench.tex':

          3,342.47 msec task-clock:u              #    0.999 CPUs utilized          
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
            71,528      page-faults:u             #    0.021 M/sec                  
    11,026,953,394      cycles:u                  #    3.299 GHz                    
    20,076,439,542      instructions:u            #    1.82  insn per cycle         
     4,054,912,588      branches:u                # 1213.148 M/sec                  
        48,023,398      branch-misses:u           #    1.18% of all branches        

       3.345258992 seconds time elapsed

       2.946411000 seconds user
       0.352941000 seconds sys
```

`perf record`
```
  22.81%  heradoc          libc-2.28.so        [.] __memmove_avx_unaligned_erms
  14.03%  heradoc          heradoc             [.] <heradoc::backend::latex::simple::TextGen as heradoc::backend::MediumCodeGenUnit<alloc::borrow::Cow<str>>>::gen
   7.72%  heradoc          heradoc             [.] alloc::string::String::push
   5.56%  heradoc          [unknown]           [k] 0xffffffff8e600163
   4.41%  heradoc          heradoc             [.] heradoc::backend::latex::replace::replace
   3.22%  heradoc          heradoc             [.] heradoc::frontend::Frontend::convert_event
   3.08%  heradoc          heradoc             [.] heradoc::generator::iter::Iter::next
```

[crdt-details.md]: https://github.com/xi-editor/xi-editor/blob/564cbde392a2ba8d78a069bcf3dc6aa6a14c2aae/docs/docs/crdt-details.md

**Comment by oberien** [comment]
Benchmark with ThinLTO (in addition to BufWriter) on 1300× [crdt-details.md] (96MB):

tldr: Slight performance increase, but compilation time increases drastically (24s → 107s, 300% increase). Worth it for actual release artifacts, but not for benchmark testing. It definitely shows that `TextGen` with its replace is one of the first spots we should take a look at. Maybe we should also reduce the `memmove`s, but their origin is harder to detect.

`perf stat ./target/release/heradoc bench.md -o bench.tex`
```
 Performance counter stats for './target/release/heradoc bench.md -o bench.tex':

          3,007.98 msec task-clock:u              #    0.999 CPUs utilized          
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
            71,497      page-faults:u             #    0.024 M/sec                  
    10,215,185,112      cycles:u                  #    3.396 GHz                    
    19,689,622,039      instructions:u            #    1.93  insn per cycle         
     3,880,948,709      branches:u                # 1290.216 M/sec                  
        46,634,502      branch-misses:u           #    1.20% of all branches        

       3.010357961 seconds time elapsed

       2.684792000 seconds user
       0.319228000 seconds sys
```

`perf record`
```
  22.58%  heradoc          libc-2.28.so        [.] __memmove_avx_unaligned_erms
  13.34%  heradoc          heradoc             [.] <heradoc::backend::latex::simple::TextGen as heradoc::backend::MediumCodeGenUnit<alloc::borrow::Cow<str>>>::gen
   9.23%  heradoc          heradoc             [.] alloc::string::String::push
   5.43%  heradoc          [unknown]           [k] 0xffffffff8e600163
   4.39%  heradoc          heradoc             [.] heradoc::backend::latex::replace::replace
   3.50%  heradoc          heradoc             [.] <heradoc::frontend::concat::Concat as core::iter::traits::iterator::Iterator>::next
   3.23%  heradoc          heradoc             [.] pulldown_cmark::parse::FirstPass::parse_line
```

[crdt-details.md]: https://github.com/xi-editor/xi-editor/blob/564cbde392a2ba8d78a069bcf3dc6aa6a14c2aae/docs/docs/crdt-details.md
